### PR TITLE
Bind provider search property to ssl object

### PR DESCRIPTION
**Pull request by wbeck10**
The SSL_CTX is created with the library context which can contain multiple providers.
At this stage the SSL_CTX object contains the provider query parameter and all the settings for SSL connection.

It makes sense for developer to create a SSL_CTX early at the application entry and apply all the settings before creating the SSL object. However, it can be the case that the app needs to toggle between provider preference when creating SSL objects.
This pull request will add the possibility to bind the provider query parameter to the SSL object by using a new SSL_new_ex API.
During the SSL connection, the provider query parameter from the SSL object will be used when set otherwise the SSL_CTX provider query parameter will be used (which is the current way).

### CODE REVIEW

File: ssl/ssl_rsa.c [hunk 1]
```diff
@@ -473,7 +475,11 @@ static int use_certificate_chain_file(SSL_CTX *ctx, SSL *ssl, const char *file)
         goto end;
     }
 
-    x = X509_new_ex(real_ctx->libctx, real_ctx->propq);
+    propq = real_ctx->propq;
+    if (ssl && ssl->propq)
```
Comment (paulidale) [inline ssl/ssl_rsa.c:479]: Please include the `!= NULL` for each as per coding style.

File: ssl/t1_trce.c [hunk 2]
```diff
@@ -1328,7 +1329,10 @@ static int ssl_print_raw_public_key(BIO *bio, const SSL *ssl, int server,
     BIO_indent(bio, indent, 80);
     BIO_printf(bio, "raw_public_key, length=%d\n", (int)clen);
 
-    pkey = d2i_PUBKEY_ex(NULL, &msg, clen, ssl->ctx->libctx, ssl->ctx->propq);
+    propq = ssl->ctx->propq;
+    if (ssl && ssl->propq)
```
Comment (paulidale) [inline ssl/t1_trce.c:1333]: Please include the `!= NULL` for each as per style guide.

File: ssl/ssl_lib.c [hunk 3]
```diff
@@ -694,6 +694,26 @@ int SSL_CTX_set_ssl_version(SSL_CTX *ctx, const SSL_METHOD *meth)
...
+        goto err;
+
+    if (propq != NULL) {
+        ret->propq = OPENSSL_strdup(propq);
+        if (ret->propq == NULL)
+            goto err;
+    }
+    return ret;
+err:
+    if (ret)
```
Comment (paulidale) [inline ssl/ssl_lib.c:712]: Just drop this, SSL_free handles NULL correctly.

File: ssl/ssl_local.h [hunk 4]
```diff
@@ -1815,6 +1816,13 @@ struct ssl_connection_st {
     SSL_CONNECTION_FROM_SSL_ONLY_int(ssl, const)
 # define SSL_CONNECTION_GET_CTX(sc) ((sc)->ssl.ctx)
 # define SSL_CONNECTION_GET_SSL(sc) (&(sc)->ssl)
+# define SSL_CONNECTION_PROV_QUERRY(spq) \
+      (spq != NULL ? \
+          (SSL_CONNECTION_GET_SSL(spq) && SSL_CONNECTION_GET_SSL(spq)->propq != NULL ? \
+          SSL_CONNECTION_GET_SSL(spq)->propq : SSL_CONNECTION_GET_CTX(spq)->propq) : NULL)
+# define SSL_PROV_QUERRY(s) \
+      (s != NULL ? (s->propq ? s->propq : s->ctx->propq) : NULL)
```
Comment (paulidale) [inline ssl/ssl_local.h:1824]: Missing `!= NULL`

File: ssl/ssl_local.h [hunk 5]
```diff
@@ -1815,6 +1816,13 @@ struct ssl_connection_st {
     SSL_CONNECTION_FROM_SSL_ONLY_int(ssl, const)
 # define SSL_CONNECTION_GET_CTX(sc) ((sc)->ssl.ctx)
 # define SSL_CONNECTION_GET_SSL(sc) (&(sc)->ssl)
+# define SSL_CONNECTION_PROV_QUERRY(spq) \
+      (spq != NULL ? \
+          (SSL_CONNECTION_GET_SSL(spq) && SSL_CONNECTION_GET_SSL(spq)->propq != NULL ? \
+          SSL_CONNECTION_GET_SSL(spq)->propq : SSL_CONNECTION_GET_CTX(spq)->propq) : NULL)
```
Comment (paulidale) [inline ssl/ssl_local.h:1822]: The indention here is quite off and it's fairly complicated.
Maybe consider making this a function and spreading everything nicely?

File: ssl/ssl_lib.c [hunk 6]
```diff
@@ -694,6 +694,26 @@ int SSL_CTX_set_ssl_version(SSL_CTX *ctx, const SSL_METHOD *meth)
 }
 #endif
 
+SSL *SSL_new_ex(SSL_CTX *ctx, const char* propq)
```
Comment (paulidale) [inline ssl/ssl_lib.c:697]: Is this better than introducing another call to set this?
That's how a lot of the other SSL settings that also exist in the SSL_CTX are implemented.

**Review by paulidale — CHANGES_REQUESTED** [review]

**Comment by paulidale** [comment]
This needs documentation and possibly a test case.

**Comment by wbeck10** [comment]
> This needs documentation and possibly a test case.

Documentation and test case added.

File: test/sslapitest.c [hunk 7]
```diff
@@ -6999,6 +6999,56 @@ static int test_ssl_clear(int idx)
...
+    SSL_CTX *cctx = NULL, *sctx = NULL;
+    SSL *clientssl = NULL, *serverssl = NULL;
+    int testresult = 0;
+
+    /* Create an initial connection */
+    if (!TEST_true(create_ssl_ctx_pair(libctx, TLS_server_method(),
+                                       TLS_client_method(), TLS1_VERSION, 0,
+                                       &sctx, &cctx, cert, privkey))
+            || !TEST_true(SSL_CTX_set_max_proto_version(cctx,
+                                                        TLS1_2_VERSION)))
```
Comment (t8m) [inline test/sslapitest.c:7013]: Why is the test limiting the TLS versions?

File: test/sslapitest.c [hunk 8]
```diff
@@ -11419,6 +11469,7 @@ int setup_tests(void)
     ADD_ALL_TESTS(test_key_update_peer_in_read, 2);
     ADD_ALL_TESTS(test_key_update_local_in_write, 2);
     ADD_ALL_TESTS(test_key_update_local_in_read, 2);
+    ADD_ALL_TESTS(test_ssl_new_ex, 2);
```
Comment (t8m) [inline test/sslapitest.c:11472]: If the test did not limit the TLS versions it could be outside any ifndefs

File: ssl/ssl_lib.c [hunk 9]
```diff
@@ -3029,6 +3049,20 @@ long ossl_ctrl_internal(SSL *s, int cmd, long larg, void *parg, int no_quic)
...
+{
+    /* Return propq from SSL object prior when set,
+     * otherwise use the SSL_CTX propq.
+     */
+    if (ssl_conn != NULL) {
+        if (SSL_CONNECTION_GET_SSL(ssl_conn) != NULL)
+            return SSL_CONNECTION_GET_SSL(ssl_conn)->propq;
+        else if (SSL_CONNECTION_GET_CTX(ssl_conn) != NULL)
+            return SSL_CONNECTION_GET_CTX(ssl_conn)->propq;
+    }
```
Comment (t8m) [inline ssl/ssl_lib.c:3062]: I do not think it does what you comment here about. `SSL_CONNECTION_GET_SSL()` basically cannot return NULL unless ssl_conn is NULL.

Also the multiline comment must start with `\*` on a separate line.

File: ssl/ssl_rsa.c [hunk 10]
```diff
@@ -442,6 +443,7 @@ static int use_certificate_chain_file(SSL_CTX *ctx, SSL *ssl, const char *file)
     pem_password_cb *passwd_callback;
     void *passwd_callback_userdata;
     SSL_CTX *real_ctx = (ssl == NULL) ? ctx : ssl->ctx;
+    const char* propq = NULL;
```
Comment (t8m) [inline ssl/ssl_rsa.c:446]: Nit: space before `*` not after

**Review by t8m — CHANGES_REQUESTED** [review]
I did not do a complete review yet - these are from the first look.

**Comment by mattcaswell** [comment]
I question the premise behind this PR.

There are numerous things pre-fetched and cached in the `SSL_CTX`. This means that some things will have been fetched using the propq associated with the `SSL_CTX` and somethings will be fetched using the propq from the `SSL`. This is quite confusing for any developer.

I think over time we should potentially look at doing more of that to the point where there are no fetches at all done at the `SSL` level. By adding this API we actually restrict our future options for caching more things in the SSL_CTX. Suddenly such a change would become "breaking" for applications that relied on this API.

I am placing on OTC hold on this.

**Comment by mattcaswell** [comment]
**OTC Question: Do we want this API?**

**Comment by wbeck10** [comment]
The reason for this PR is to used different provider preference when creating an SSL connection.
The SSL_CTX has a lot of parameters to be set. For a developer, it is not convenient to cache all the settings and then create a SSL_CTX (with preferred provider query) shortly before creating a connection. It is more convenient to create a global SSL_CTX and set directly the parameter to it. It is quite possible that the underlying lib ctx provider store contains multiple providers. I believe, that some user may want to choose the preferred provider per SSL connection.

To solve this problem, this PR binds the provider query to the SSL object. If the SSL provider query is set, then it will use this when fetching a digest or cipher, otherwise it would be the provider query from the SSL_CTX. 
 
There is no API available to clone a SSL_CTX (I believe), besides, I think the SSL_CTX takes quite a lot of memory as well.  Maybe the ssl_st structure is not the right place to put provider query in, please suggest how it can be done and if defining the provider query per SSL connection makes sense or not.

**Comment by paulidale** [comment]
This PR isn't achieving what you want.  As @mattcaswell noted, lots of things are prefetched and cached at the SSL_CTX level.  This includes all the TLS related algorithms e.g.  Subsequently setting the property query on the SSL object is not going to regenerate this cache, so it will be the same as if you were fetching using the old query.

**Comment by wbeck10** [comment]
Is it maybe possible to split the configuration part of an SSL_CTX into a new structure and allow SSL_CTX to be generated with the new configuration structure?

**Comment by t8m** [comment]
I could also see SSL_CTX_dup being implemented if that would help. Given refcounting it probably would not be memory-intensive that much.

**Comment by t8m** [comment]
**OTC: This API cannot be added as is.**

As SSL_CTX can contain fetched algorithms, keys, or other things that depend on property query, it would be inconsistent to try to override the propq on SSL_new(). Could you please give us more information about the architecture of the application where having separate SSL_CTX objects for separate propqueries is infeasible.

**Comment by ifranzki** [comment]
While reading this another idea came to my mind. Not sure if this would help @wbeck10 with what he is trying to solve though....

I think it would be beneficial if one could create an SSL context with not just one property query argument.
Currently, `SSL_CTX_new_ex()` takes exactly one property query, which is then used for every crypto operation that is done with this SSL context.

There are use cases, where one might want to use different providers for the different operations that are performed during SSL/TLS handshake and data exchange.

For example, I would like to use a HSM/PKCS#11 based provider for the signature operation during TLS handshake, but would like to use the default provider for the key exchange operations and the data encryption. 

Rational behind this: The private key of the TLS server (or even client) is the identity of the server, and thus should be HSM protected. I would like to use an HSM/PKCS#11 based key for that, so the signature operations would need to use a HSM/PKCS#11 based provider. However, for the ephemeral ECDH key exchange and the symmetric session key a clear key implementations (i.e. the default provider) is good enough. The short lived (ephemeral) keys do not need to be HSM protected. 

Currently, with just one property query for the SSL context, when we use an EC based cipher suite with ECDSA for signatures, and ECDHE key exchange, it will most likely use the same provider for ECDSA and ECDH.

Keep in mind that doing a HSM/PKCS#11 based ECDH is very hard (or even impossible) because a HSM/PKCS#11 based ECDH operation with a non-extractable EC private key typically results in a non-extractable HSM/PKCS#11 based ECDH symmetric key object as well. So without the support for opaque symmetric keys (aka SKEY - discussed elsewhere already), one needs to do tricks to be able to use a HSM/PKCS#11 provider to secure a TLS handshake with a HSM/PKCS#11 based ECDSA step but a clear key ECDH key exchange step. 

I could think of and `SSL_CTX_new_ex2()` function that takes 2  (or even more) property queries. One for the signature operations, another one for the key exchange operations, and maybe a 3rd for the symmetric encryption operations. 

If the SSL context caches the providers used, then I guess it should need to cache separate instances for signature, key exchange and symmetric (possibly a 4th one for digests, etc....)

This is of curse not fully thought through, but I guess you get the idea.

**Comment by wbeck10** [comment]
Hi t8m,

Sorry for the late reply, I was in holiday.
The reason we need the provider property bind to SSL object is following:
Our application can have different provider added before establishing a SSL connection.
In our case it is an internal developed CNG (Windows trust store) provider and a P11 provider.
Both providers are added in the library context.
After that, a global SSL_CTX is generated to be used as an object that can store all the SSL_CTX settings like cipher suites, timeout and so forth.
When trying to make an SSL connection, it is known at this point that we want to use for example the CNG provider but unfortunately the provider parameter is defined in the SSL_CTX.

Start App
 |
\/
Create library context and add default, Cng and P11 providers
 |
\/
Create global SSL_CTX (to store SSL settings)
 |
\/
Set some SSL CTX settings
 |
\/
Create SSL object to make SSL connection


We thinking as well to create multiple SSL connection in multithreads with different provider preferences which I think would require having the SSL_CTX generated per thread (I don’t think SSL_CTX is thread safe, isn’t it?).

So the basic problem is the configuration that is currently enclosed in the SSL_CTX.
I think, if I can implement an SSL_CTX_dup and set the provider parameter for each SSL_CTX duplicate, that would work for us.  

Any thought?

**Comment by wbeck10** [comment]
Another pull request is created to go with the SSL_CTX_dup way.
(https://github.com/openssl/openssl/pull/22395)

**Pull request closed without merging**